doc: steps and info to setup envtest (#3510)
operator-framework/operator-sdk: website/content/en/docs/building-operators/golang/references/env-test-setup.md

---
title: EnvTest Setup
linkTitle: EnvTest Setup
weight: 50
---

## Overview 

> This document describes how to configure the environment for the [controller tests][controller-test] supported by the SDK.

Reviewer: Link to [envtest](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/envtest) somewhere in the overview.

## Configuring your test 

The suite test requires specify the `kubectl`, `api-server` and `etcd` the k8s binaries which are used by it. You can use the following script as a helper which will create the `testbin/` directory with these binaries. Following an example by adding a new target in your Makefile. 

```sh
# Setup binaries required to run the tests
# The script will create the testbin dir and add on it the required binaries

> # See that it expects the Kubernetes and ETCD version
> testsetup:
> 	curl -sSLo setup_envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/kubebuilder/master/scripts/setup_envtest_bins.sh 
> 	chmod +x setup_envtest.sh
> 	./setup_envtest.sh v1.18.2 v3.4.3

Reviewer: To make what these versions refer to obvious, wrap them in env vars:
```suggestion
# See that it expects the Kubernetes and ETCD version
K8S_VERSION = v1.18.2
ETCD_VERSION = v3.4.3
testbin:
	curl -sSLo setup_envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/kubebuilder/master/scripts/setup_envtest_bins.sh 
	chmod +x setup_envtest.sh
	./setup_envtest.sh $(K8S_VERSION) $(ETCD_VERSION)
```

```

Also, note that the above script will use environment variables to specify the place where its binaries can be found:

```shell

>   export TEST_ASSET_KUBECTL=<kubectl bin path>
>   export TEST_ASSET_KUBE_APISERVER=<api-server bin path>
>   export TEST_ASSET_ETCD=/<etd bin path>

Reviewer: ```suggestion
$ export TEST_ASSET_KUBECTL=<kubectl-bin-path>
$ export TEST_ASSET_KUBE_APISERVER=<api-server-bin-path>
$ export TEST_ASSET_ETCD=<etcd-bin-path>
```

---
title: EnvTest Setup
linkTitle: EnvTest Setup
weight: 50
---

## Overview 

This document describes how to configure the environment for the [controller tests][controller-test] which uses [envtest][envtest] and is supported by the SDK. 

## Installing prerequisites

> [Envtest][envtest] requires that the `kubectl`, `api-server` and `etcd` be present locally. You can use this [script][script] to download these binaries into the `testbin/` directory. It may be convenient to add this script your Makefile as follows:

Reviewer 2: ```suggestion
[Envtest][envtest] requires that `kubectl`, `api-server` and `etcd` be present locally. You can use this [script][script] to download these binaries into the `testbin/` directory. It may be convenient to add this script your Makefile as follows:
```

```sh
# Setup binaries required to run the tests
# See that it expects the Kubernetes and ETCD version
K8S_VERSION = v1.18.2
ETCD_VERSION = v3.4.3
testbin:
	curl -sSLo setup_envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/kubebuilder/master/scripts/setup_envtest_bins.sh 
	chmod +x setup_envtest.sh
	./setup_envtest.sh $(K8S_VERSION) $(ETCD_VERSION)
```

> The above script sets these environment variables to specify where test binaries can be found:

Reviewer 3: If the script is the thing setting the environment variables, wouldn't we have to `source` the script to pick up the environment variables once that script has completed?

Author: The source is the testbin/ dir which is created in the project. I will try to clarifies that.

Author: yes, I am just trying to clarify here how it works in order to allow users to perform their setup as they wish. I will try to improve that.

> The above script sets these environment variables to specify where test binaries can be found. Then, see that if you would like to NOT use the script you can use the same configuration by informing the path of your binaries: 

Author: ```suggestion
The above script sets these environment variables to specify where test binaries can be found. In case you would like to not use the script then, is possible to do the same configuration to inform the path of your binaries: 
```
